src/contexts/ExchangeIssuance/ExchangeIssuanceProvider.tsx

```diff
@@ -0,0 +1,250 @@
+import useBalances from 'hooks/useBalances'
+import useTransactionWatcher from 'hooks/useTransactionWatcher'
+import { useDebounce } from 'hooks/useDebounce'
+import { getIssuanceTradeTransaction } from 'index-sdk/exchangeIssuance'
+import {
+  getIssuanceTradeType,
+  getIssuanceCallData,
+  getIssuanceTransactionOptions,
+} from './utils'
+import trackReferral from 'utils/referralApi'
+import { waitTransaction } from 'utils/index'
+import { TransactionStatusType } from 'contexts/TransactionWatcher'
+import { currencyTokens } from 'constants/currencyTokens'
+import { IssuancePriceData } from './types'
+
+const ExchangeIssuanceProvider: React.FC = ({ children }) => {
+  const [issuanceToken, setIssuanceToken] = useState<'dpi' | 'index' | 'cgi'>(
+    'dpi'
+  )
+  const [isFetchingOrderData, setIsFetchingOrderData] = useState<boolean>(false)
+  const [isUserIssuing, setIsUserIssuing] = useState<boolean>(true)
+  const [activeField, setActiveField] = useState<'currency' | 'set'>('currency')
+  const [selectedCurrency, setSelectedCurrency] = useState<any>()
+  const [currencyQuantity, setCurrencyQuantity] = useState<string>()
+  const [tokenQuantity, setTokenQuantity] = useState<string>()
+  const [currencyOptions, setCurrencyOptions] = useState<any[]>([])
+  const [issuanceData, setIssuanceData] = useState<IssuancePriceData>(
+    {} as IssuancePriceData
+  )
+
+  const { onSetTransactionId, onSetTransactionStatus } = useTransactionWatcher()
+
+  const {
+    ethBalance,
+    dpiBalance,
+    cgiBalance,
+    indexBalance,
+    daiBalance,
+    usdcBalance,
+  } = useBalances()
+
+  const {
+    account,
+    ethereum,
+  }: { account: string | null | undefined; ethereum: provider } = useWallet()
+
+  useEffect(() => {
+    setCurrencyOptions(currencyTokens)
+    setSelectedCurrency(currencyTokens[0])
+  }, [])
+
+  let spendingTokenBalance = new BigNumber(0)
+  if (!isUserIssuing && issuanceToken === 'index') {
+    spendingTokenBalance = indexBalance || new BigNumber(0)
+  } else if (!isUserIssuing && issuanceToken === 'dpi') {
+    spendingTokenBalance = dpiBalance || new BigNumber(0)
+  } else if (!isUserIssuing && issuanceToken === 'cgi') {
+    spendingTokenBalance = cgiBalance || new BigNumber(0)
+  } else if (selectedCurrency?.id === 'wrapped_eth') {
+    spendingTokenBalance = ethBalance || new BigNumber(0)
+  } else if (selectedCurrency?.id === 'mcd') {
+    spendingTokenBalance = daiBalance || new BigNumber(0)
+  } else if (selectedCurrency?.id === 'usdc') {
+    spendingTokenBalance = usdcBalance || new BigNumber(0)
+  }
+
+  const debouncedCurrencyQuantity = useDebounce(currencyQuantity)
+  const debouncedTokenQuantity = useDebounce(tokenQuantity)
+  const targetTradeQuantity =
+    activeField === 'currency'
+      ? debouncedCurrencyQuantity
+      : debouncedTokenQuantity
+
+  useEffect(() => {
+    if (!targetTradeQuantity) return
+
+    setIsFetchingOrderData(true)
+    fetchTokenBuySellData(
```

Comment:
You shouldn't be using any tokensets api to fetch buy/sell data. these should be contract calls to exchange issuance contract

Reply:
Check out this section for getting token estimations
https://docs.google.com/document/d/1ayfDW4mGoQoWEhuI8DsOlR-czO6zjs10p1Zefh6VmW8/edit#heading=h.pckzbf2w3v8n

Reply:
resolved